tests/telemetry/events/run_events_test.py

```diff
@@ -3,26 +3,34 @@
 import json
 import logging
 import os
+import re
 
-
+from tests.common.utilities import wait_until
 from telemetry_utils import listen_for_events
 logger = logging.getLogger(__name__)
 
 
 def run_test(duthost, gnxi_path, ptfhost, data_dir, validate_yang, trigger, json_file,
-             filter_event_regex, tag):
+             filter_event_regex, tag, heartbeat=False, thread_timeout=30):
     op_file = os.path.join(data_dir, json_file)
-    heartbeat = trigger is None
-    if not heartbeat:  # no trigger for heartbeat
+    if trigger is not None:  # no trigger for heartbeat
         trigger(duthost)  # add events to cache
-    listen_for_events(duthost, gnxi_path, ptfhost, filter_event_regex, op_file)  # listen from cache
+    listen_for_events(duthost, gnxi_path, ptfhost, filter_event_regex, op_file,
+                      thread_timeout)  # listen from cache
     data = {}
     with open(op_file, "r") as f:
         data = json.load(f)
     assert len(data) > 0, "Did not parse regex from output: {}".format(filter_event_regex)
     logger.info("events received: ({})".format(json.dumps(data, indent=4)))
     if heartbeat:  # no yang validation for heartbeat
         return
-    dest = "/tmp/" + json_file
+    dest = "~/" + json_file
     duthost.copy(src=op_file, dest=dest)
     validate_yang(duthost, dest, tag)
+    wait_until(5, 1, 0, is_gnmi_cli_finished, duthost)
+
+
+def is_gnmi_cli_finished(duthost):
+    last_logs = duthost.shell("tail -n 2 /var/log/syslog | grep eventd", module_ignore_errors=True)["stdout"]
```

Comment:
>tail -n 2 [](http://example.com/codeflow?start=31&length=9)

If you repeatedly call `tail` for last 2 lines, you may miss some logging if there are lots coming.

Reply:
Now check for log lines in the last 2 seconds repeatedly

Reply:
sed with exact timestamp will have issue if that timestamp has no matching line.

Reply:
Changed back to tail -n 2. I think it's okay if logs are missed as we are waiting for eventd to turn caching on which takes around 1 second. We can check syslog to confirm that before moving on to next event. If we miss due to excessive logs, the wait_until will provide a long enough buffer anway for eventd to reset for next test.

Reply:
Added further explanation inside PR description